src/haywire/http_request_buffers.h

```diff
@@ -0,0 +1,81 @@
+#pragma once
+#include <stdbool.h>
+#include "khash.h"
+#include "haywire.h"
+
+typedef struct {
+    void* buffer;
+    size_t size;
+} hw_request_buffer_chunk;
+
+typedef void* hw_request_buffer;
+
+/**
+ * Initializes a new buffer. The caller is responsible for calling http_request_buffer_destroy to free up memory.
+ */
+hw_request_buffer* http_request_buffer_init(size_t max_size);
+
+/**
+ * Signals that "size" bytes from the buffer are now in use.
+ */
+void http_request_buffer_consume(hw_request_buffer* buffer, size_t size);
+
+/**
+ * Marks all buffer chunks up to the last one allocated (exclusive) for removal when
+ * http_request_buffer_sweep gets called.
+ */
+void http_request_buffer_mark(hw_request_buffer* buffer);
+
+/**
+ * Sweeps all buffer chunks up to the mark. Data for the last chunk is copied to the beginning of the buffer
+ * and the offsets are updates accordingly.
+ */
+void http_request_buffer_sweep(hw_request_buffer* buffer);
+
+/**
+ * Prints the buffer.
+ */
+void http_request_buffer_print(hw_request_buffer* buffer);
+
+/**
+ * Ensures that there's will be a sizable chunk available when the it's requested via http_request_buffer_chunk.
+ * This will ensure that the requested size doesn't exceed the maximum buffer size and will try to
+ * re-use the existing underlying buffers if there's sufficient space still to be used.
+ * Therefore, "requested_size" is just a hint, and not necessarily the size that will be allocated.
+ *
+ * Returns true if the allocation succeeds, false otherwise (errno is set accordingly).
+ */
+bool http_request_buffer_alloc(hw_request_buffer* buf, size_t requested_size);
+
+/**
+ * Returns a new buffer chunk to be used by request reader.
+ */
+void http_request_buffer_chunk(hw_request_buffer* buffer, hw_request_buffer_chunk* chunk);
+
+/**
+ * Informs the buffer manager that *pointer is a memory region of interest and that will have to be made available to
+ * the caller eventually, when http_request_buffer_locate is called. The pin is given an id by the caller so it can be
+ * retrieved later. Pins will be overwritten if the same id is used multiple times.
+ */
+void http_request_buffer_pin(hw_request_buffer* buffer, unsigned long id, void* pointer);
```

Comment:
As we discussed on IRC perhaps `unsigned long id` should be `void* id` or would `void* key` be more intuitive?